### PR TITLE
Simplify exponentiation in SMT backend

### LINKED ISSUE #99: SMT: further simplify exponentiation expressions
The SMT backend currently performs some very simple preprocessing of expressions involving exponentation (see [here](https://github.com/ethereum/act/blob/3944f4f06b73ee75b7226128b2bd01676f1ea21b/src/SMT.hs#L410)). These could be extended to allow the SMT encoding of any exponentiation where the RHS is concrete, by either:

- fully evaluating the result of the exponentation if both the RHS & LHS are concrete
- encoding the exponentiation as a chain of multiplications if the RHS is concrete, but the LHS is symbolic

We will need to write a simple evaluator for act expressions to do this properly.

### PR DESCRIPTION
**Pull request by kjekac**
Closes #99, i.e. we can now pass any exponentiation where the exponent is concrete to the SMT backend even though it has no support for exponentiation. This does not however include any algebraic analysis to attempt to simplify away symbolic values.

I'm a bit uncertain on how division is to be dealt with. Currently I'm using `div` and `mod` from Haskell's Prelude (rather than `quot` and `rem`). This is because they truncate towards negative infinity, which is what the EVM seems to be doing according to the yellow paper. But what exactly are we trying to agree with? The EVM? In that case `x/0` should be `0`, so we need a special case for that. Or Solidity? Or our other backends?

### CODE REVIEW

File: src/RefinedAst.hs [hunk 1]
```diff
@@ -215,6 +220,51 @@ data ReturnExp
   | ExpBytes  (Exp ByteString)
   deriving (Eq, Show)
 
+-- | Simplifies concrete expressions into literals.
+-- Returns `Nothing` if the expression contains symbols.
+eval :: Exp a -> Maybe a
+eval e = case e of
+  And  a b    -> [a' && b' | a' <- eval a, b' <- eval b]
```
Comment (MrChico) [inline src/RefinedAst.hs:227]: monad comprehension! Cool, haven't seen this before

**Comment by kjekac** [comment]
> > But what exactly are we trying to agree with? The EVM? In that case x/0 should be 0, so we need a special case for that. Or Solidity? Or our other backends?
> 
> IMO, specs should conform to "mathematical rigor", and not so much evm details, meaning I think we should ideally disallow specs which admit division by zero as part of future spec validity checking (discussed at #14).

Ok so following that reasoning we want to truncate downwards instead of towards zero, since that's how Euclidean division works. Then the code here should be correct.

File: src/SMT.hs [hunk 2]
```diff
@@ -406,10 +409,12 @@ expToSMT2 ctx@(Ctx behvName whn) e = case e of
 
 -- | SMT2 has no support for exponentiation, but we can do some preprocessing
 --   if the RHS is concrete to provide some limited support for exponentiation
---   TODO: support any exponentiation expression where the RHS evaluates to a concrete value
 simplifyExponentiation :: Exp Integer -> Exp Integer -> Exp Integer
-simplifyExponentiation (LitInt a) (LitInt b) = LitInt (a ^ b)
-simplifyExponentiation _ _ = error "Internal Error: exponentiation is unsupported in SMT lib"
+simplifyExponentiation a b = fromMaybe (error "Internal Error: no support for symbolic exponents in SMT lib")
+                             $   [LitInt $ a' ^ b'                      | a' <- eval a,   b' <- evalb]
+                             <|> [foldr Mul (LitInt 1) (replicate b' a) | b' <- fromInteger <$> evalb]
```
Comment (xwvvvvwx) [inline src/SMT.hs:415]: What will happen with the `fromInteger` here if the exponent is larger than the bounds of the `Int` type?

Comment (xwvvvvwx) [reply src/SMT.hs:415]: seems like it will just return 0?

```
Prelude> let a = (2 ^ 256) :: Integer
Prelude> (fromInteger a) :: Int
0
```

Comment (xwvvvvwx) [reply src/SMT.hs:415]: definately an edge case, but I would prefer to either keep everything as an `Integer` here or hard fail if we overflow the `Int` type.

**Review by xwvvvvwx — APPROVED** [review]

**Pull request merged**